source/opt/def_use_manager.h

```diff
@@ -32,54 +32,70 @@
+struct Use {
+  ir::Instruction* inst;   // Instruction using the id.
+  uint32_t operand_index;  // logical operand index of the id use. This can be
+                           // the index of result type id.
+};
+
+using UseList = std::list<Use>;
+
 // A class for analyzing and managing defs and uses in an ir::Module.
 class DefUseManager {
  public:
-  using InstLocPair = std::pair<ir::Instruction*, uint32_t>;
   using IdToDefMap = std::unordered_map<uint32_t, ir::Instruction*>;
-  using IdToUseMap = std::unordered_map<uint32_t, std::list<InstLocPair>>;
+  using IdToUsesMap = std::unordered_map<uint32_t, UseList>;
 
   // Analyzes the defs and uses in the given |module| and populates data
   // structures in this class.
+  // TODO(antiagainst): This method should not modify the given module. Create
+  // const overload for ForEachInst().
   void AnalyzeDefUse(ir::Module* module);
 
   // Returns the def instruction for the given |id|. If there is no instruction
   // defining |id|, returns nullptr.
   ir::Instruction* GetDef(uint32_t id);
   // Returns the use instructions for the given |id|. If there is no uses of
   // |id|, returns nullptr.
-  std::list<InstLocPair>* GetUses(uint32_t id);
+  UseList* GetUses(uint32_t id);
 
   // Returns the map from ids to their def instructions.
-  const IdToDefMap& id_to_defs() const { return id_to_defs_; }
+  const IdToDefMap& id_to_defs() const { return id_to_def_; }
   // Returns the map from ids to their uses in instructions.
-  const IdToUseMap& id_to_uses() const { return id_to_uses_; }
+  const IdToUsesMap& id_to_uses() const { return id_to_uses_; }
 
   // Turns the instruction defining the given |id| into a Nop. Returns true on
-  // success, false if the given |id| is not defined at all. This method does
-  // not erase the uses of |id|.
+  // success, false if the given |id| is not defined at all. This method also
+  // erases both the uses of |id| and the |id|-generating instruction's use
+  // information kept in this manager, but not the operands in the original
+  // instructions.
   bool KillDef(uint32_t id);
   // Replaces all uses of |before| id with |after| id. Returns true if any
   // replacement happens. This method does not kill the definition of the
-  // |before| id.
+  // |before| id. If |after| is the same as |before|, does nothing and returns
+  // false.
   bool ReplaceAllUsesWith(uint32_t before, uint32_t after);
 
  private:
+  using IdToParamMap = std::unordered_map<uint32_t, std::vector<uint32_t>>;
+
   // Analyzes the defs and uses in the given |inst|.
   void AnalyzeInstDefUse(ir::Instruction* inst);
 
-  IdToDefMap id_to_defs_;  // Mapping from ids to their definitions
-  IdToUseMap id_to_uses_;  // Mapping from ids to their uses
+  IdToDefMap id_to_def_;       // Mapping from ids to their definitions
+  IdToUsesMap id_to_uses_;     // Mapping from ids to their uses
+  IdToParamMap id_to_params_;  // Mapping from ids to those ids used as operands
```

Comment:
Hm.  Now that I look at the code to build this, it's a map from the Ids of instructions that use an id, to the list of ids they use, and it can contain duplicates.
Better to call this id_to_used_ids_ or at least have a better comment.